File: 0229-majority-element-ii/0229-majority-element-ii.java
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        List<Integer> res = new ArrayList<>();
        int ct1 = 0, ct2 = 0; // atmost 2 elements can be there
        int el1 = 0, el2 = 0; // Moores algo
        for(int i=0; i< nums.length; i++){
            if(ct1 == 0 && el2 != nums[i]){  // if we dont check el2 != nums[i]
            // then a case can occur that el1 and el2 have same value
                ct1++;
                el1 = nums[i];
            } 
            else if(ct2 == 0 && el1 != nums[i]){// if we dont check el1 != nums[i]
            // then a case can occur that el1 and el2 have same value
                ct2++;
                el2 = nums[i];
            } 
            else if(nums[i] == el1)
                ct1++;
            else if(nums[i] == el2)
                ct2++;
            else{
                ct1--; // other element canceling el1 count
                ct2--; // some other element canceled el2 count
            }
        }
        System.out.println(el1 + " " + el2);
        ct1 =0; ct2 =0; 
        for(int i=0; i< nums.length; i++){
            if(nums[i] == el1)
                ct1++;
            else if(nums[i] == el2)
                ct2++;
        }
        int minreq = (nums.length/3) +1; // this is the required count 
        if(ct1 >= minreq)
            res.add(el1);
        if(ct2 >= minreq)
            res.add(el2);
        return res;
    }
}

File: 0229-majority-element-ii/0229-majority-element-ii.py
from typing import List

class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        res = [] # There can be at most two majority elements which are more than ⌊n/3⌋ times
        ct1, ct2 = 0, 0
        el1, el2 = 0, 1  # el1 and el2 must be different to avoid conflict

        # First pass to find potential candidates
        for num in nums:
            if ct1 == 0 and num != el2: # num != el2 is important as number need to be different
                ct1 = 1
                el1 = num # new candidate for el1 as ct1 was 0
            elif ct2 == 0 and num != el1:
                ct2 = 1
                el2 = num # new candidate for el2 as ct2 was 0
            elif num == el1:
                ct1 += 1
            elif num == el2:
                ct2 += 1
            else: # when num doesnt match both then decrease both counts
                ct1 -= 1
                ct2 -= 1

        # Second pass to verify the counts
        ct1, ct2 = 0, 0
        for num in nums:
            if num == el1:
                ct1 += 1
            elif num == el2:
                ct2 += 1

        minreq = len(nums) // 3 + 1
        if ct1 >= minreq:
            res.append(el1)
        if ct2 >= minreq:
            res.append(el2)

        return res
